File: src/lib/historyStats.ts
import { addDaysToIsoDate, isValidSessionCalendarDate } from "./sessionCalendar";

export const TRAILING_4_WEEK_DAYS = 28;
const SECONDS_IN_4_WEEKS = TRAILING_4_WEEK_DAYS * 24 * 60 * 60;
const TEN_THOUSAND_HOURS_SECONDS = 10_000 * 60 * 60;

export type HistorySessionTimeInput = {
  sessionDate?: string;
  actualTime?: number | null;
  duration: number;
  bonusSeconds?: number;
};

export type HistoryPeriodStats = {
  trailing4WeekDays: number;
  trailing4WeekDayPercent: number;
  trailing4WeekTotalTime: number;
  trailing4WeekTimePercent: number;
  totalTimeAllTime: number;
  progressTo10kHours: number;
};

/** Seconds counted toward history time totals (elapsed wall time for the sit). */
export function sessionTimeSeconds(session: HistorySessionTimeInput): number {
  return session.actualTime ?? session.duration;
}

/**
 * Trailing 4-week and all-time stats for the History view (#83).
 * `todayIso` should match the client's local calendar day (same convention as
 * `sessionDate` stamping) so the window ends on the user's "today".
 */
export function calculateHistoryPeriodStats(
  sessions: HistorySessionTimeInput[],
  todayIso: string,
): HistoryPeriodStats {
  const periodStart = addDaysToIsoDate(todayIso, -(TRAILING_4_WEEK_DAYS - 1));
  const datesInPeriod = new Set<string>();
  let trailing4WeekTotalTime = 0;
  let totalTimeAllTime = 0;

  for (const session of sessions) {
    if (!isValidSessionCalendarDate(session.sessionDate)) continue;
    const secs = sessionTimeSeconds(session);
    totalTimeAllTime += secs;
    if (session.sessionDate! >= periodStart && session.sessionDate! <= todayIso) {
      datesInPeriod.add(session.sessionDate!);
      trailing4WeekTotalTime += secs;
    }
  }

  const trailing4WeekDays = datesInPeriod.size;
  const trailing4WeekDayPercent = parseFloat(
    ((trailing4WeekDays / TRAILING_4_WEEK_DAYS) * 100).toFixed(1),
  );
  const trailing4WeekTimePercent = parseFloat(
    ((trailing4WeekTotalTime / SECONDS_IN_4_WEEKS) * 100).toFixed(2),
  );
  const progressTo10kHours = parseFloat(
    ((totalTimeAllTime / TEN_THOUSAND_HOURS_SECONDS) * 100).toFixed(2),
  );

  return {
    trailing4WeekDays,
    trailing4WeekDayPercent,
    trailing4WeekTotalTime,
    trailing4WeekTimePercent,
    totalTimeAllTime,
    progressTo10kHours,
  };
}

File: ios/StillPointShared/Sources/StillPointShared/HistoryStats.swift
import Foundation

public enum HistoryStats {
    public static let trailing4WeekDays = 28
    private static let secondsIn4Weeks = trailing4WeekDays * 24 * 60 * 60
    private static let tenThousandHoursSeconds = 10_000 * 60 * 60

    public struct PeriodStats: Sendable, Equatable {
        public let trailing4WeekDays: Int
        public let trailing4WeekDayPercent: Double
        public let trailing4WeekTotalTime: Int
        public let trailing4WeekTimePercent: Double
        public let totalTimeAllTime: Int
        public let progressTo10kHours: Double

        public init(
            trailing4WeekDays: Int,
            trailing4WeekDayPercent: Double,
            trailing4WeekTotalTime: Int,
            trailing4WeekTimePercent: Double,
            totalTimeAllTime: Int,
            progressTo10kHours: Double
        ) {
            self.trailing4WeekDays = trailing4WeekDays
            self.trailing4WeekDayPercent = trailing4WeekDayPercent
            self.trailing4WeekTotalTime = trailing4WeekTotalTime
            self.trailing4WeekTimePercent = trailing4WeekTimePercent
            self.totalTimeAllTime = totalTimeAllTime
            self.progressTo10kHours = progressTo10kHours
        }
    }

    /// Trailing 4-week and all-time stats for the History view (#83).
    /// `todayIso` should match the client's local calendar day (same convention as
    /// `sessionDate` stamping) so the window ends on the user's "today".
    public static func calculatePeriodStats(
        sessions: [some HistorySessionTimeInput],
        todayIso: String
    ) -> PeriodStats {
        let periodStart = SessionCalendar.addDays(toIsoDate: todayIso, deltaDays: -(trailing4WeekDays - 1))
        var datesInPeriod = Set<String>()
        var trailing4WeekTotalTime = 0
        var totalTimeAllTime = 0

        for session in sessions {
            guard SessionCalendar.isValidSessionCalendarDate(session.sessionDate) else { continue }
            let secs = HistorySessionTime.sessionTimeSeconds(session)
            totalTimeAllTime += secs
            if session.sessionDate >= periodStart && session.sessionDate <= todayIso {
                datesInPeriod.insert(session.sessionDate)
                trailing4WeekTotalTime += secs
            }
        }

        let trailing4WeekDayCount = datesInPeriod.count
        let trailing4WeekDayPercent = roundToOneDecimal(
            (Double(trailing4WeekDayCount) / Double(trailing4WeekDays)) * 100
        )
        let trailing4WeekTimePercent = roundToTwoDecimals(
            (Double(trailing4WeekTotalTime) / Double(secondsIn4Weeks)) * 100
        )
        let progressTo10kHours = roundToTwoDecimals(
            (Double(totalTimeAllTime) / Double(tenThousandHoursSeconds)) * 100
        )

        return PeriodStats(
            trailing4WeekDays: trailing4WeekDayCount,
            trailing4WeekDayPercent: trailing4WeekDayPercent,
            trailing4WeekTotalTime: trailing4WeekTotalTime,
            trailing4WeekTimePercent: trailing4WeekTimePercent,
            totalTimeAllTime: totalTimeAllTime,
            progressTo10kHours: progressTo10kHours
        )
    }

    private static func roundToOneDecimal(_ value: Double) -> Double {
        (value * 10).rounded() / 10
    }

    private static func roundToTwoDecimals(_ value: Double) -> Double {
        (value * 100).rounded() / 100
    }
}
